Factor dynamic linking section into its own file

dschuff opened the pull request:
Fix up some references about dynamic linking and move it to its own file, since
it will be expanding soon and referring to a fair number of other sections.
Also add a little extra motivation to the introductory text.

lukewagner commented on Modules.md:
```diff
@@ -102,7 +102,7 @@ scope of "all" is a nuanced question), a single app using multiple
 transitively used by those libraries "played well" together (e.g., explicitly
 shared `malloc` and coordinated global address ranges). Instead, the
-[dynamic linking future feature](FutureFeatures.md#dynamic-linking) is intended
+[dynamic linking future feature](DynamicLinking.md) is intended
```
I think there are a fair number more references to FutureFeatures.md#dynamic-linking (you can run `tools/generate.py` to check link validity).

jfbastien replied:
About that... [It doesn't work at the moment](https://github.com/WebAssembly/design/blob/master/tools/generate.py#L128) because *someone* left a `TODO`. Sorry!

dschuff replied:
It wouldn't have helped anyway, since there's still a dynamic linking section in FutureFeatures. I'll find and update the rest of them.

lukewagner commented on DyamicLinking.md:
```diff
@@ -0,0 +1,34 @@
+
+WebAssembly will support both load-time and run-time (`dlopen`) dynamic linking
+of both WebAssembly modules and non-WebAssembly modules (e.g., on the Web, ES6
+ones containing JavaScript).
```
Wow, I see this is pre-existing, but I think this para is totally wrong: we've only ever discussed dynamically linking wasm modules; I'm not even sure what dynamically linking an ES6 module would mean.

dschuff replied:
Yeah, I wondered about that. We will already have module imports and exports, (so load-time linking), but in general it's probably outside the scope of this design anyway, so I'll just remove it.

jfbastien commented on DyamicLinking.md:
```diff
@@ -0,0 +1,34 @@
+specified ABI for interoperability, WebAssembly runtimes will be ABI agnostic,
+so it will be possible to use a non-standard ABI for specialized purposes.
+
+Although dynamic linking is not part of the MVP, it is important and has
```
Drop "is important and".

dschuff replied:
Done.

jfbastien commented on DyamicLinking.md:
```diff
@@ -0,0 +1,34 @@
+such as the way linear memory is managed, how module imports and exports are
+specified, and how globals and function pointers work.  Therefore we want to
+have some viable ideas to ensure we don't standardize a design that
+unnecessarily complicates the design or implication of dynamic linking.
```
Implementation?

dschuff replied:
Done.

dschuff commented:
Updated, PTAL. I also removed language referring to dynamically-linked libraries as "modules" since that has important distinctions in this context. I'll probably just use the term "dynamic libraries" (or dylibs for short?) instead of DSOs since this is independent of whether libraries are actually shared. pre-bikeshedding on that welcome now, or whenever :)

jfbastien commented:
lgtm
